sctest/comparator: modify line differently for LSC and DSC

Xiang-Gu opened the pull request:
Previously, in the comparator testing framework, we attempt to modify
the input line if it's one of the known cases that would cause behavioral
difference between LSC and DSC. One such example is ALTER PK with rowid
as the old primary key column, in which case we would rewrite the stmt
to something like
```
ALTER PK; COMMIT; DROP COLUMN rowid
```
so that the LSC matches the behavior of the DSC where ALTER PK would
already have dropped the rowid column. The oversight, however, is that
executing such a multi-statement-line would use LSC only! This defeats
the purpose of applying the modification in the first place. To address
this, this commit changes the modification step so that it now returns
different lines for the LSC and for the DSC to execute. On the same
example above, we would now get
```
LSC: ALTER PK; COMMIT; DROP COLUMN rowid;  -- unchanged
DSC: ALTER PK;
```

Informs #113351

Xiang-Gu committed:
sctest: modify line differently for LSC and DSC

Previously, in the comparator testing framework, we attempt to modify
the input line if it's one of the known cases that would cause behavioral
difference between LSC and DSC. One such example is ALTER PK with rowid
as the old primary key column, in which case we would rewrite the stmt
to something like
```
ALTER PK; COMMIT; DROP COLUMN rowid
```
so that the LSC matches the behavior of the DSC where ALTER PK would
already have dropped the rowid column. The oversight, however, is that
executing such a multi-statement-line would use LSC only! This defeats
the purpose of applying the modification in the first place. To address
this, this commit changes the modification step so that it now returns
different lines for the LSC and for the DSC to execute. On the same
example above, we would now get
```
LSC: ALTER PK; COMMIT; DROP COLUMN rowid;  -- unchanged
DSC: ALTER PK;
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114441)
<!-- Reviewable:end -->

chrisseto commented on pkg/sql/schemachanger/sctest/comparator.go:
```diff
@@ -74,25 +74,22 @@ func CompareLegacyAndDeclarative(t *testing.T, ss StmtLineReader) {
 		// code.
 		_, errLegacy := legacyConn.ExecContext(ctx, line)
-		if pgcode.MakeCode(string(getPQErrCode(errLegacy))) == pgcode.FeatureNotSupported && !containsCommit(line) {
+		if pgcode.MakeCode(string(getPQErrCode(errLegacy))) == pgcode.FeatureNotSupported &&
```
This if statement is starting to get a bit difficult to understand. How about making some intermediate variables or adding some helpers?

```go
//  Is this trying to tell if legacy has been modified from the original or if declarative and legacy differ regardless of modifications?
legacyLineModified := line != declarativeLine
// Also uncertain what this is attempting to detect
IDontKnowWhatToCallThis := legacyLineModified || !containsCommit(line)

if isFeatureNotSupported(errLegacy) && IDontKnowWhatToCallThis {
    continue
}
```

chrisseto commented on pkg/sql/schemachanger/sctest/comparator.go:
```diff
@@ -74,25 +74,22 @@ func CompareLegacyAndDeclarative(t *testing.T, ss StmtLineReader) {
-		}
+		// descriptor state. This step is primarily for accounting for the known
+		// behavior difference between the two schema changers.
+		line, declarativeLine = preExecutionProcessing(ctx, t, line, legacyConn, declarativeConn)
```
This is starting to get pretty complex and difficult to follow as a 3rd party. It may be worthwhile to add in a struct to help wrap up all these variables and provide a convenient place for documentation.

```go
// Line represents a single SQL (line | statement) that is to be executed and cross examined across the legacy and declarative schema changers
type Line struct {
Original string

// Legacy is a copy of Original, potentially with modifications, for execution on the Legacy Schema Changer.
Legacy string

// You get the idea ...
Declarative string
}
```

chrisseto commented on pkg/sql/schemachanger/sctest/comparator.go:
```diff
@@ -163,32 +152,77 @@ func isInATransaction(ctx context.Context, t *testing.T, conn *gosql.Conn) bool
 	}
-	return parsedLine.StringWithFlags(tree.FmtSimple | tree.FmtTagDollarQuotes)
+	// `EXPLAIN(DDL,SHAPE) stmt` returns an error if statement is not supported in the DSC.
+	_, err := conn.ExecContext(ctx, fmt.Sprintf("EXPLAIN(DDL, SHAPE) %v", line))
```
nit: use `%s` for `Sprintf`ing strings

chrisseto commented on pkg/sql/schemachanger/sctest/comparator.go:
```diff
@@ -163,32 +152,77 @@ func isInATransaction(ctx context.Context, t *testing.T, conn *gosql.Conn) bool
+	}
+
+	parsedLineForLegacy, err := parser.Parse(line)
 	require.NoError(t, err)
```
Isn't this error case the same as `hasSyntaxError`?

chrisseto commented on pkg/sql/schemachanger/sctest/comparator.go:
```diff
@@ -163,32 +152,77 @@ func isInATransaction(ctx context.Context, t *testing.T, conn *gosql.Conn) bool
+		}
+	}
+
 	if modify {
```
Have you considered always logging all 3 variants of `line` rather than tracking modified? You could branch on `testing.Verbose()` if you wanted a way to silence the logs.

chrisseto commented on pkg/sql/schemachanger/sctest/comparator.go:
```diff
@@ -163,32 +152,77 @@ func isInATransaction(ctx context.Context, t *testing.T, conn *gosql.Conn) bool
+	return legacyLine, declarativeLine
+}
+
+func willLineBeExecutedInDSC(
```
I'm a bit confused about the need for this function. Couldn't the DSC modifiers perform a type switch to determine if a change is needed?

chrisseto reviewed (approved):
Nothing blocking but It's starting to get a bit difficult to follow the code as is. Applying a bit more structure could go a long way for maintainability IMO

Xiang-Gu committed:
sctest/comparator: Refactor and make line modification cleaner

This is pure mechanical and non-functional change where we refactor and
make the logic for line modification and logging in the comparator
testing framework leaner and cleaner.

Release note: None

Xiang-Gu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114441)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto)

___
*[`pkg/sql/schemachanger/sctest/comparator.go` line 83 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114441#gh-1401273188:-NkMFFp07Yxq6aq3UCcX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf4da7c8219bf5e0d69c68d7fc113789be5d5970/pkg/sql/schemachanger/sctest/comparator.go#L83)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

This is starting to get pretty complex and difficult to follow as a 3rd party. It may be worthwhile to add in a struct to help wrap up all these variables and provide a convenient place for documentation.

```go
// Line represents a single SQL (line | statement) that is to be executed and cross examined across the legacy and declarative schema changers
type Line struct {
Original string

// Legacy is a copy of Original, potentially with modifications, for execution on the Legacy Schema Changer.
Legacy string

// You get the idea ...
Declarative string
}
```
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/sctest/comparator.go` line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114441#gh-1401266665:-NkMHqAZBMDKpGzbli9p:bcl10zj) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf4da7c8219bf5e0d69c68d7fc113789be5d5970/pkg/sql/schemachanger/sctest/comparator.go#L88)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

This if statement is starting to get a bit difficult to understand. How about making some intermediate variables or adding some helpers?

```go
//  Is this trying to tell if legacy has been modified from the original or if declarative and legacy differ regardless of modifications?
legacyLineModified := line != declarativeLine
// Also uncertain what this is attempting to detect
IDontKnowWhatToCallThis := legacyLineModified || !containsCommit(line)

if isFeatureNotSupported(errLegacy) && IDontKnowWhatToCallThis {
    continue
}
```
</blockquote></details>

done. In fact, with the introduction of the `Line` structure, we can simplify this check condition!
___
*[`pkg/sql/schemachanger/sctest/comparator.go` line 170 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114441#gh-1401283324:-NkMHzt-8sSiPWGxc7bt:bmw025f) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf4da7c8219bf5e0d69c68d7fc113789be5d5970/pkg/sql/schemachanger/sctest/comparator.go#L170)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Isn't this error case the same as `hasSyntaxError`?
</blockquote></details>

yes, with some refactoring, I've deleted `hasSyntaxError`
___
*[`pkg/sql/schemachanger/sctest/comparator.go` line 208 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114441#gh-1401284511:-NkMLv6_5Ae9DRQ8lGk6:bmavtzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf4da7c8219bf5e0d69c68d7fc113789be5d5970/pkg/sql/schemachanger/sctest/comparator.go#L208)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Have you considered always logging all 3 variants of `line` rather than tracking modified? You could branch on `testing.Verbose()` if you wanted a way to silence the logs.
</blockquote></details>

I decided to instead log only `Original` if no modification and `Original, Legacy, Declarative` if modification.
___
*[`pkg/sql/schemachanger/sctest/comparator.go` line 217 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114441#gh-1401285366:-NkMMPzzERuo_D1RMHTi:b96ll04) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf4da7c8219bf5e0d69c68d7fc113789be5d5970/pkg/sql/schemachanger/sctest/comparator.go#L217)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

I'm a bit confused about the need for this function. Couldn't the DSC modifiers perform a type switch to determine if a change is needed?
</blockquote></details>

1. My intention here is "only perform the modification, which was intended to account for known behavioral difference between LSC and DSC, if the line is indeed going to be executed in DSC. Otherwise, we don't bother doing any modification at all since this line will be executed under LSC in both clusters".
2. Unfortunately, it's not enough to use a type switch to determine whether a line will be executed in DSC, because a line might contain multiple statements (e.g. line = "alter table add column; alter another table drop column;"). In addition, even if line is a single statement line (e.g. line = "alter table add column;"), we still have fallback cases for DDL statements that we "claim" to support, so one way to accurately know whether a single statement line is supported in DSC is to run `EXPLAIN(DDL)` (or its friends) on the line, which is a explain feature specific for DSC. 


<!-- Sent from Reviewable.io -->

Xiang-Gu commented:
TFTR!

bors r+
